From a8970d810acc6ebf3345f5dc0ed5005a03416ab6 Mon Sep 17 00:00:00 2001
From: charliechen <charliechen@rk-forge>
Date: Mon, 15 Jun 2026 11:44:01 +0800
Subject: [PATCH] ARM: dts: rockchip: add rk3506 SoC + AES RK3506B board DT

Minimal rk3506 SoC DT (pinctrl+clk already upstream) + board DT for
ATK RK3506B (board id 'aes').

- rk3506.dtsi: minimal SoC nodes (cru/cpus/psci/timer/gic/grf+ioc syscons/
  uart0@ff0a0000/otp/pinctrl+5 gpio banks); earlycon console at 0xff0a0000
- rk3506b-aes.dts: board (model 'AES RK3506B Board', memory 512M, &uart0 okay,
  chosen stdout-path serial0:1500000n8 + earlycon)
- Makefile: build rk3506b-aes.dtb

Boots mainline Linux 7.0.12 to userspace handoff on real RK3506B
(vendor SPL/DDR/TEE + mainline U-Boot loading zImage+dtb from SPI NAND).

Signed-off-by: charliechen <charliechen@rk-forge>
---
 arch/arm/boot/dts/rockchip/Makefile        |   3 +-
 arch/arm/boot/dts/rockchip/rk3506.dtsi     | 280 +++++++++++++++++++++
 arch/arm/boot/dts/rockchip/rk3506b-aes.dts |  28 +++
 3 files changed, 310 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm/boot/dts/rockchip/rk3506.dtsi
 create mode 100644 arch/arm/boot/dts/rockchip/rk3506b-aes.dts

diff --git a/arch/arm/boot/dts/rockchip/Makefile b/arch/arm/boot/dts/rockchip/Makefile
index 716f5540e..3b76d23ea 100644
--- a/arch/arm/boot/dts/rockchip/Makefile
+++ b/arch/arm/boot/dts/rockchip/Makefile
@@ -43,4 +43,5 @@ dtb-$(CONFIG_ARCH_ROCKCHIP) += \
 	rk3288-veyron-pinky.dtb \
 	rk3288-veyron-speedy.dtb \
 	rk3288-veyron-tiger.dtb \
-	rk3288-vyasa.dtb
+	rk3288-vyasa.dtb \
+	rk3506b-aes.dtb
diff --git a/arch/arm/boot/dts/rockchip/rk3506.dtsi b/arch/arm/boot/dts/rockchip/rk3506.dtsi
new file mode 100644
index 000000000..64a1bd026
--- /dev/null
+++ b/arch/arm/boot/dts/rockchip/rk3506.dtsi
@@ -0,0 +1,280 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * (C) Copyright 2023 Rockchip Electronics Co., Ltd.
+ *
+ * Minimal RK3506 SoC dtsi for mainline Linux bring-up (earlycon/console milestone).
+ * Ported/reduced from the ATK vendor SDK rk3502.dtsi. Only the nodes needed to
+ * reach the early UART console are included: clocks, cru, grf, ioc syscons,
+ * gic, arch timer, psci, cpus, pinctrl (all 5 gpio banks) + uart0.
+ * All other peripherals (i2c/spi/pwm/sai/mmc/usb/display/...) deferred —
+ * they are not required for earlycon. Add them in later iterations.
+ */
+
+#include <dt-bindings/clock/rockchip,rk3506-cru.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <dt-bindings/pinctrl/rockchip.h>
+
+/ {
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	compatible = "rockchip,rk3506";
+
+	interrupt-parent = <&gic>;
+
+	aliases {
+		gpio0 = &gpio0;
+		gpio1 = &gpio1;
+		gpio2 = &gpio2;
+		gpio3 = &gpio3;
+		gpio4 = &gpio4;
+		serial0 = &uart0;
+	};
+
+	clocks {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		clk_rc: clk-rc {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <400000>;
+			clock-output-names = "clk_rc";
+		};
+
+		xin24m: xin24m {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <24000000>;
+			clock-output-names = "xin24m";
+		};
+
+		xin32k: xin32k {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <32768>;
+			clock-output-names = "xin32k";
+		};
+	};
+
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu0: cpu@f00 {
+			device_type = "cpu";
+			compatible = "arm,cortex-a7";
+			reg = <0xf00>;
+			enable-method = "psci";
+			clocks = <&cru ARMCLK>;
+		};
+
+		cpu1: cpu@f01 {
+			device_type = "cpu";
+			compatible = "arm,cortex-a7";
+			reg = <0xf01>;
+			enable-method = "psci";
+			clocks = <&cru ARMCLK>;
+		};
+
+		cpu2: cpu@f02 {
+			device_type = "cpu";
+			compatible = "arm,cortex-a7";
+			reg = <0xf02>;
+			enable-method = "psci";
+			clocks = <&cru ARMCLK>;
+		};
+	};
+
+	psci {
+		compatible = "arm,psci-1.0";
+		method = "smc";
+	};
+
+	timer {
+		compatible = "arm,armv7-timer";
+		interrupts = <GIC_PPI 13 (GIC_CPU_MASK_SIMPLE(3) | IRQ_TYPE_LEVEL_HIGH)>,
+			     <GIC_PPI 14 (GIC_CPU_MASK_SIMPLE(3) | IRQ_TYPE_LEVEL_HIGH)>,
+			     <GIC_PPI 11 (GIC_CPU_MASK_SIMPLE(3) | IRQ_TYPE_LEVEL_HIGH)>,
+			     <GIC_PPI 10 (GIC_CPU_MASK_SIMPLE(3) | IRQ_TYPE_LEVEL_HIGH)>;
+		clock-frequency = <24000000>;
+	};
+
+	uart0: serial@ff0a0000 {
+		compatible = "rockchip,rk3506-uart", "snps,dw-apb-uart";
+		reg = <0xff0a0000 0x100>;
+		interrupts = <GIC_SPI 34 IRQ_TYPE_LEVEL_HIGH>;
+		reg-shift = <2>;
+		reg-io-width = <4>;
+		clocks = <&cru SCLK_UART0>, <&cru PCLK_UART0>;
+		clock-names = "baudclk", "apb_pclk";
+		pinctrl-names = "default";
+		pinctrl-0 = <&uart0_xfer_pins>;
+		status = "disabled";
+	};
+
+	grf: syscon@ff288000 {
+		compatible = "rockchip,rk3506-grf", "syscon", "simple-mfd";
+		reg = <0xff288000 0x4000>;
+	};
+
+	ioc_grf: syscon@ff4d8000 {
+		compatible = "rockchip,rk3506-ioc-grf", "syscon";
+		reg = <0xff4d8000 0x8000>;
+	};
+
+	gic: interrupt-controller@ff581000 {
+		compatible = "arm,gic-400";
+		reg = <0xff581000 0x1000>,
+		      <0xff582000 0x2000>,
+		      <0xff584000 0x2000>,
+		      <0xff586000 0x2000>;
+		interrupts = <GIC_PPI 9 IRQ_TYPE_LEVEL_HIGH>;
+		#interrupt-cells = <3>;
+		interrupt-controller;
+		#address-cells = <0>;
+	};
+
+	ioc1: syscon@ff660000 {
+		compatible = "rockchip,rk3506-ioc1", "syscon";
+		reg = <0xff660000 0x10000>;
+	};
+
+	grf_pmu: syscon@ff910000 {
+		compatible = "rockchip,rk3506-grf-pmu", "syscon", "simple-mfd";
+		reg = <0xff910000 0x4000>;
+	};
+
+	ioc_pmu: syscon@ff950000 {
+		compatible = "rockchip,rk3506-ioc-pmu", "syscon";
+		reg = <0xff950000 0x10000>;
+	};
+
+	cru: clock-controller@ff9a0000 {
+		compatible = "rockchip,rk3506-cru";
+		reg = <0xff9a0000 0x20000>;
+		rockchip,grf = <&grf>;
+		#clock-cells = <1>;
+		#reset-cells = <1>;
+
+		assigned-clocks =
+			<&cru CLK_32K_FRAC>,
+			<&cru CLK_FRAC_UART_MATRIX0>,
+			<&cru CLK_FRAC_UART_MATRIX1>;
+		assigned-clock-rates =
+			<32768>,
+			<96000000>,
+			<128000000>;
+	};
+
+	pinctrl: pinctrl {
+		compatible = "rockchip,rk3506-pinctrl";
+		rockchip,grf = <&ioc_grf>;
+		rockchip,ioc1 = <&ioc1>;
+		rockchip,pmu = <&ioc_pmu>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		pcfg_pull_up: pcfg-pull-up {
+			bias-pull-up;
+		};
+
+		pcfg_pull_none: pcfg-pull-none {
+			bias-disable;
+		};
+
+		gpio0: gpio@ff940000 {
+			compatible = "rockchip,gpio-bank";
+			reg = <0xff940000 0x200>;
+			interrupts = <GIC_SPI 0 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 1 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 2 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 3 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cru PCLK_GPIO0>, <&cru DBCLK_GPIO0>;
+
+			gpio-controller;
+			#gpio-cells = <2>;
+			gpio-ranges = <&pinctrl 0 0 32>;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+		};
+
+		gpio1: gpio@ff870000 {
+			compatible = "rockchip,gpio-bank";
+			reg = <0xff870000 0x200>;
+			interrupts = <GIC_SPI 4 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 5 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 6 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 7 IRQ_TYPE_LEVEL_HIGH>;
+
+			clocks = <&cru PCLK_GPIO1>, <&cru DBCLK_GPIO1>;
+
+			gpio-controller;
+			#gpio-cells = <2>;
+			gpio-ranges = <&pinctrl 0 32 32>;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+		};
+
+		gpio2: gpio@ff1c0000 {
+			compatible = "rockchip,gpio-bank";
+			reg = <0xff1c0000 0x200>;
+			interrupts = <GIC_SPI 8 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 9 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 10 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 11 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cru PCLK_GPIO2>, <&cru DBCLK_GPIO2>;
+
+			gpio-controller;
+			#gpio-cells = <2>;
+			gpio-ranges = <&pinctrl 0 64 32>;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+		};
+
+		gpio3: gpio@ff1d0000 {
+			compatible = "rockchip,gpio-bank";
+			reg = <0xff1d0000 0x200>;
+			interrupts = <GIC_SPI 12 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 13 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 14 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 15 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cru PCLK_GPIO3>, <&cru DBCLK_GPIO3>;
+
+			gpio-controller;
+			#gpio-cells = <2>;
+			gpio-ranges = <&pinctrl 0 96 32>;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+		};
+
+		gpio4: gpio@ff1e0000 {
+			compatible = "rockchip,gpio-bank";
+			reg = <0xff1e0000 0x200>;
+			interrupts = <GIC_SPI 16 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 17 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 18 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 19 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cru PCLK_GPIO4>, <&cru DBCLK_GPIO4>;
+
+			gpio-controller;
+			#gpio-cells = <2>;
+			gpio-ranges = <&pinctrl 0 128 32>;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+		};
+
+		uart0 {
+			uart0_xfer_pins: uart0-xfer-pins {
+				rockchip,pins =
+					/* uart0_rx */ <0 RK_PC7 1 &pcfg_pull_up>,
+					/* uart0_tx */ <0 RK_PC6 1 &pcfg_pull_up>;
+			};
+		};
+	};
+};
diff --git a/arch/arm/boot/dts/rockchip/rk3506b-aes.dts b/arch/arm/boot/dts/rockchip/rk3506b-aes.dts
new file mode 100644
index 000000000..0d45ae455
--- /dev/null
+++ b/arch/arm/boot/dts/rockchip/rk3506b-aes.dts
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Board device tree for the AES RK3506B board (RK3506B SoC, arm32 Cortex-A7).
+ * Mainline bring-up: minimal board for earlycon/console on uart0.
+ */
+
+/dts-v1/;
+
+#include "rk3506.dtsi"
+
+/ {
+	model = "AES RK3506B Board";
+	compatible = "aes,rk3506b-aes", "rockchip,rk3506b", "rockchip,rk3506";
+
+	chosen {
+		stdout-path = "serial0:1500000n8";
+		bootargs = "earlycon=uart8250,mmio32,0xff0a0000 console=ttyS0";
+	};
+
+	memory@0 {
+		device_type = "memory";
+		reg = <0x0 0x20000000>;	/* 512 MiB; DRAM base 0x0 (CFG_SYS_SDRAM_BASE=0) */
+	};
+};
+
+&uart0 {
+	status = "okay";
+};
-- 
2.43.0

